Controls+

CubesterYT opened the pull request:
Hello all! It's been a while since this extension had been in development, for 4 months if I remember, but here it is, the finished product, Controls+!

Here's the roster:
![image](https://github.com/TurboWarp/extensions/assets/78769806/cba483d8-4c46-4810-984e-cf3073de1ab6)

This is a finished product, but I would happily accept any suggestions to add.

I would like to ask you guys if I should rename these blocks:
![image](https://github.com/TurboWarp/extensions/assets/78769806/2e2bb22c-17ca-43c9-b8d4-0005c4039bea)

I'm not sure about those titles, but if you guys have better suggestions for those, let me know!

TODO:
 - Add documentation
 - Change banner
 
 👍 reactions help a lot!

CubesterYT committed:
Controls+ (Note, current banner is a placeholder)

SharkPool-SP commented:
You're controlling :trollface:

CubesterYT commented:
@SharkPool-SP, Lmao, on god bro

Ashimee commented:
forever if is just a while loop, your execute and run case blocks have seemingly no difference, all of the switch case blocks break in the palette and your outer loop function is buggy, also why is there pre existing blocks in here

CubesterYT committed:
Added two new blocks, removed wait frames or until, and used negative branch method for wait seconds or until.

CubesterYT commented:
**UPDATE**
Added two new blocks:
`for each [VARIABLE] in [LIST]`, and `for each [VARIABLE] in [LIST] from [VALUE]`
![image](https://github.com/TurboWarp/extensions/assets/78769806/3e139daf-c97d-46b1-a416-7be0b09337bc)
Every loop, it'll set the variable to the corresponding item number from the chosen list.

Removed `wait [FRAMES] frames or until [CONDITIONAL`

Shortened the name of `for each [VARIABLE] in [TIMES] starting from [VALUE]` to `for each [VARIABLE] in [TIMES] from [VALUE]`

Used the negative branch method for the `wait [SECONDS] seconds or until [CONDITIONAL]` block to make it work.

CubesterYT committed:
Rearranged two blocks, and made `run green flag` terminal.

CubesterYT committed:
Some sources

CubesterYT committed:
Formatting, bruh

CubesterYT committed:
Rework the menus, wanted to make it even simpler, but it didn't update on demand.

CubesterYT committed:
Add `restart script` block

CubesterYT commented:
# UPDATE

New block: `restart script`

![image](https://github.com/TurboWarp/extensions/assets/78769806/ac739a9e-9e2f-441e-a847-618020bea93e)

This block, when run, will fetch the entire script it's in and restart it.

CubesterYT committed:
Add some new icons, and also utilize the new `branchIconURI` option!

CubesterYT commented:
**UPDATE**

Added some new icons and utilized the new `branchIconURI` option!
![image](https://github.com/TurboWarp/extensions/assets/78769806/444a179b-b032-4d02-b4af-7a2f6ec0561f)

CubesterYT committed:
Added `do... while [CONDITIONAL]`

CubesterYT commented:
**UPDATE**

Added a new block called `do... while [CONDITIONAL]`
![image](https://github.com/TurboWarp/extensions/assets/78769806/1b150ef5-0dfe-4d08-beaf-57cd44af41fa)

This block runs the code within the loop at least ***once***, and it the conditional is true, it'll keep looping till it is false. If it is false after the first run, it'll continue the rest of the script.

CubesterYT committed:
Major Oopsie :skull:

CubesterYT committed:
Fix Switch Case blocks breaking in the palette!

CubesterYT committed:
Massive Update

CubesterYT commented:
# Major Update
This update is pretty big, here's what we have:
 - A good rewrite of the Switch Case function to fix every bug, it works perfectly now. Thanks UnLuckyPanda on Discord for notifying me about the bugs.
 - Addition of the `(switch value)` reporter. (finally)
 - Removal of both `forever if` blocks. They were too niche.
 - Added `color3` to complete the Controls color scheme
 - Utilized the new `extensions` option for blocks to sync the block colors to the Controls Category. ***
 - Utilized `branchIconURI: ""` to make the `wait [SECONDS] seconds or until [CONDITIONAL]` block look normal. ***
 - Rearranged some blocks.
 

  *** (Currently works in the merge upstream test and will work normally when merge upstream is officially added)

CubesterYT committed:
Quick efficiency update

CubesterYT committed:
Added `whileScript` loop, this is going to be godly

CubesterYT commented:
## Update
![image](https://github.com/TurboWarp/extensions/assets/78769806/1c925a79-fa3e-4812-8bdc-ec0be02fb292)

```js
while
  script
run
  script
```
This block, while the first script runs, will loop the second script until the first one finishes! (I'm going to be using this block so much brooo)

JeremyGamer13 commented:
> **UPDATE**
> 
> Added some new icons and utilized the new `branchIconURI` option! ![image](https://private-user-images.githubusercontent.com/78769806/292780418-444a179b-b032-4d02-b4af-7a2f6ec0561f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQyNzA1ODIsIm5iZiI6MTcwNDI3MDI4MiwicGF0aCI6Ii83ODc2OTgwNi8yOTI3ODA0MTgtNDQ0YTE3OWItYjAzMi00ZDAyLWI0YWYtN2EyZjZlYzA1NjFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTAzVDA4MjQ0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxZDU1YjAyNDVhM2U1ZTUxMmViNGU2MjBkOGUzYzhlNTFlMzlhYmJlZmY3OTI1ZDFjNWJlMjk1ZjQxODMzMjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WdNNEyNE2Y7Mvq-B6k7T0ZvsSNYZ_8RGGBrfV1_C5Pk)

is this still the icon because the outline on the arrow looks a bit weird

CubesterYT commented:
> > **UPDATE**
> > 
> > Added some new icons and utilized the new `branchIconURI` option! ![image](https://private-user-images.githubusercontent.com/78769806/292780418-444a179b-b032-4d02-b4af-7a2f6ec0561f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQyNzA1ODIsIm5iZiI6MTcwNDI3MDI4MiwicGF0aCI6Ii83ODc2OTgwNi8yOTI3ODA0MTgtNDQ0YTE3OWItYjAzMi00ZDAyLWI0YWYtN2EyZjZlYzA1NjFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTAzVDA4MjQ0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxZDU1YjAyNDVhM2U1ZTUxMmViNGU2MjBkOGUzYzhlNTFlMzlhYmJlZmY3OTI1ZDFjNWJlMjk1ZjQxODMzMjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WdNNEyNE2Y7Mvq-B6k7T0ZvsSNYZ_8RGGBrfV1_C5Pk)
> 
> is this still the icon because the outline on the arrow looks a bit weird


@JeremyGamer13, yes, it is still the icon?

GreexHead commented:
Doesn't the `do... while [CONDITIONAL]` block already exist in Turbowarp and Scratch as `while [CONDITIONAL]`?

CubesterYT commented:
> Doesn't the `do... while [CONDITIONAL]` block already exist in Turbowarp and Scratch as `while [CONDITIONAL]`?

@GreexHead, Nope, this block will have a guaranteed chance to run the script. It'll run the script first, then checks if it should loop.

GreexHead commented:
> > Doesn't the `do... while [CONDITIONAL]` block already exist in Turbowarp and Scratch as `while [CONDITIONAL]`?
> 
> @GreexHead, Nope, this block will have a guaranteed chance to run the script. It'll run the script first, then checks if it should loop. While checks if it should loop, and then decide loop or not.

Oh, thanks for the correction!

CubesterYT commented:
> > > Doesn't the `do... while [CONDITIONAL]` block already exist in Turbowarp and Scratch as `while [CONDITIONAL]`?
> > 
> > @GreexHead, Nope, this block will have a guaranteed chance to run the script. It'll run the script first, then checks if it should loop. While checks if it should loop, and then decide loop or not.
> 
> Oh, thanks for the correction!

@GreexHead, no problem!

GreexHead commented:
Oh yeah, one more thing...
Can you please add a `[STATE] loop` terminal block? It has two options: `break` and `continue`.

CubesterYT commented:
@GreexHead, may you elaborate?

GreexHead commented:
> @GreexHead, may you elaborate?

I'll let Blockly documentation explain.
![image](https://github.com/TurboWarp/extensions/assets/112185687/7c28cca1-c923-472f-8589-30d0b2c7f498)

CubesterYT commented:
@GreexHead, I understand break and continue, I had tried implementing it, but it's way too hard to implement, but what is the [STATE] loop terminal block you mention?

GreexHead commented:
> @GreexHead, I understand break and continue, I had tried implementing it, but it's way too hard to implement, but what is the [STATE] loop terminal block you mention?

Said block has both break and continue as options.

CubesterYT commented:
@GreexHead, no like, what does the block do?

GreexHead commented:
okay actually scratch that
have break and continue as seperate blocks

Ashimee commented:
> okay actually scratch that have break and continue as seperate blocks

these blocks are above his current skill level > but it's way too hard to implement

CubesterYT commented:
> > okay actually scratch that have break and continue as seperate blocks
> 
> these blocks are above his current skill level
> 
> > but it's way too hard to implement

Why'd you say it like that lmao?

CubesterYT committed:
Allow dropping the if then else reporter anywhere for QOL

CubesterYT commented:
## Update

Allowed dropping `if then else` reporter anywhere for QOL
![image](https://github.com/TurboWarp/extensions/assets/78769806/df02c225-8932-4b67-9084-61650437c0d9)

ZXMushroom63 commented:
The `(if <_> then (   ) else (   ))` block seems like more of an operator block than a controls block. Along with this, multiple extensions add their own implementation of it.

ZXMushroom63 commented:
> 
> @ZXMushroom63, I added to my extension because it fits the scope perfectly, and allows dropping anywhere unlike other implements. Most other implements are utility extensions, which we are all trying to work together to split up and build in them into separate extensions.

Fair enough. It is useful to be able to drop the block in boolean inputs.

LilyMakesThings commented:
> > The `(if <_> then (   ) else (   ))` block seems like more of an operator block than a controls block. Along with this, multiple extensions add their own implementation of it.

Couplers already includes a ternary operator. If you want to let it drop into other inputs, you can just submit a pr to let it do so.

Off topic: Utilities v1 is kind of an exception to the "grahhh utilities extensions" rule. It was designed so perfectly and with such a concise yet useful mix of blocks no one hates it. It's the perfect one. The ternary operator fits so well in there.

TeknikSpace commented:
Can we have an isolated run block? as in when ran it's like running it in a clone, it isolates private variables inside from the main sprite.

SharkPool-SP commented:
> Can we have an isolated run block? as in when ran it's like running it in a clone, it isolates private variables inside from the main sprite. this'd be really cool and could possibly support turbowarp's return block, also fantastic extension! keep up the good work!

I made a block for this for my extension :trollface: 
Cubester hasn't added it to this yet
<img width="309" alt="Screenshot 2024-03-02 at 7 57 11 AM" src="https://github.com/TurboWarp/extensions/assets/139097378/2ca1b397-3415-4d3e-a43e-14b397515c9e">
